lib/common/format.h

```diff
@@ -0,0 +1,164 @@
+// Copyright (c) 2019, NVIDIA CORPORATION.  All rights reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#ifndef FORTRAN_COMMON_FORMAT_H_
+#define FORTRAN_COMMON_FORMAT_H_
+
+#include "Fortran.h"
+#include <cstring>
+
+namespace Fortran::common {
+
+struct FormatError {
+  static constexpr int maxArgLength_{25};
+
+  const char *text_;  // error text; may have one %s argument
+  char arg_[maxArgLength_ + 1];  // optional %s argument value
+  int offset_;  // offset to error marker
+  int length_;  // length of error marker
+};
+
+template<typename CHAR = char> class FormatValidator {
+public:
+  explicit FormatValidator(const CHAR *format, size_t length, IoStmtKind stmt,
+      bool warnOnNonstandardUsage, FormatError *errors, int maxErrors)
+    : format_{format}, end_{format ? format + length : nullptr}, stmt_{stmt},
```

Comment:
The initialization of `end_` makes me think that you expect `format` to sometimes be a null pointer.  Can it be null, though?

Reply:
The possibility of a null format depends on caller code.  For
  write(*,fmt='')
format is currently a non-null pointer and *format is 0.  For
 100 format ()
format is a non-null pointer and *format is '('.  Allowing for format==nullptr is just extra caution for now, pending future runtime I/O library calls to this code.  If format is a nullptr, that would generate an 'Empty format expression' error, which is what happens with the write statement above.

Reply:
It looks like it is ok to do address arithmetic on a nullptr, as long as the result is not dereferenced.  I have therefore changed the constructor to check that format is not null.